src/runtime/simulator.cc

```diff
@@ -39,29 +39,841 @@ bool ParallelConfig::is_data_parallel() const
+{
+  assert(id_to_gpu_fb_mem.find(device_id) != id_to_gpu_fb_mem.end());
+  return id_to_gpu_fb_mem[device_id];
+}
+
+int MachineModel_old::get_num_gpus()
+{
+  return num_gpus;
+}
+
+float MachineModel_old::get_intra_node_gpu_bandwidth()
+{
+  return inter_gpu_bandwidth;
+}
+
+float MachineModel_old::get_inter_node_gpu_bandwidth()
+{
+  return inter_node_bandwidth;
+}
+
+
+std::vector<Comm_device *> MachineModel_old::get_comm_path(Mem_device *src_mem, Mem_device *tar_mem)
 {
-  printf("Device: node_id(%d) gpu_id(%d) capacity(%zu)\n", node_id, gpu_id, capacity);
-  assert(type == DEVICE_GPU);
+    std::vector<Comm_device *> ret;
+    // on the same memory
+    if (src_mem->mem_type == tar_mem->mem_type and src_mem->device_id == tar_mem->device_id) {
+        return ret;
+    }
+    if (src_mem->mem_type == Mem_device::SYSTEM_MEM and tar_mem->mem_type == Mem_device::SYSTEM_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            return ret;
+        }
+        else {
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
+            ret.emplace_back(ids_to_inter_node_comm_device[device_id]);
+        }
+    }
+    else if (src_mem->mem_type == Mem_device::GPU_FB_MEM and tar_mem->mem_type == Mem_device::GPU_FB_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            int device_id = src_mem->device_id * num_gpus + tar_mem->device_id;
+            ret.emplace_back(ids_to_inter_gpu_comm_device[device_id]);
+        }
+        else {
+            ret.emplace_back(id_to_gputodram_comm_device[src_mem->device_id]);
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
+            ret.emplace_back(ids_to_inter_node_comm_device[device_id]);
+            ret.emplace_back(id_to_dramtogpu_comm_device[tar_mem->device_id]);
+        }
+    }
+    else if (src_mem->mem_type == Mem_device::SYSTEM_MEM and tar_mem->mem_type == Mem_device::GPU_FB_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            ret.emplace_back(id_to_dramtogpu_comm_device[tar_mem->device_id]);
+        }
+        else {
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
+            ret.emplace_back(ids_to_inter_node_comm_device[device_id]);
+            ret.emplace_back(id_to_dramtogpu_comm_device[tar_mem->device_id]);
+        }
+    }
+    else if (src_mem->mem_type == Mem_device::GPU_FB_MEM and tar_mem->mem_type == Mem_device::SYSTEM_MEM) {
+        if (src_mem->node_id == tar_mem->node_id) {
+            ret.emplace_back(id_to_gputodram_comm_device[src_mem->device_id]);
+        }
+        else {
+            ret.emplace_back(id_to_gputodram_comm_device[src_mem->device_id]);
+            int device_id = src_mem->node_id * num_nodes + tar_mem->node_id;
+            ret.emplace_back(ids_to_inter_node_comm_device[device_id]);
+        }
+    }
+    else {
+        printf("No path found between %s and %s\n", src_mem->name.c_str(), tar_mem->name.c_str());
```

Comment:
Should this `exit` or `assert` or something, or is the expected behavior to just return an empty list?

Reply:
fixed